flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/nn/KNN.scala

```diff
@@ -0,0 +1,353 @@
+    override def fit(
+                      instance: KNN,
+                      fitParameters: ParameterMap,
+                      input: DataSet[T]): Unit = {
+      val resultParameters = instance.parameters ++ fitParameters
+
+      require(resultParameters.get(K).isDefined, "K is needed for calculation")
+
+      val blocks = resultParameters.get(Blocks).getOrElse(input.getParallelism)
+      val partitioner = FlinkMLTools.ModuloKeyPartitioner
+      val inputAsVector = input.asInstanceOf[DataSet[FlinkVector]]
+
+      instance.trainingSet = Some(FlinkMLTools.block(inputAsVector, blocks, Some(partitioner)))
+    }
+  }
+
+  /** [[PredictDataSetOperation]] which calculates k-nearest neighbors of the given testing data
+    * set.
+    * @tparam T Subtype of [[Vector]]
+    * @return The given testing data set with k-nearest neighbors
+    */
+  implicit def predictValues[T <: FlinkVector : ClassTag : TypeInformation] = {
+    new PredictDataSetOperation[KNN, T, (FlinkVector, Array[FlinkVector])] {
+      override def predictDataSet(
+                                   instance: KNN,
+                                   predictParameters: ParameterMap,
+                                   input: DataSet[T]): DataSet[(FlinkVector,
+        Array[FlinkVector])] = {
+        val resultParameters = instance.parameters ++ predictParameters
+
+        instance.trainingSet match {
+          case Some(trainingSet) =>
+            val k = resultParameters.get(K).get
+            val blocks = resultParameters.get(Blocks).getOrElse(input.getParallelism)
+            val metric = resultParameters.get(DistanceMetric).get
+            val partitioner = FlinkMLTools.ModuloKeyPartitioner
+
+            // attach unique id for each data
+            val inputWithId: DataSet[(Long, T)] = input.zipWithUniqueId
+
+            // split data into multiple blocks
+            val inputSplit = FlinkMLTools.block(inputWithId, blocks, Some(partitioner))
+
+            val sizeHint = resultParameters.get(SizeHint)
+            val crossTuned = sizeHint match {
+              case Some(hint) if hint == CrossHint.FIRST_IS_SMALL =>
+                trainingSet.crossWithHuge(inputSplit)
+              case Some(hint) if hint == CrossHint.SECOND_IS_SMALL =>
+                trainingSet.crossWithTiny(inputSplit)
+              case _ => trainingSet.cross(inputSplit)
+            }
+
+            // join input and training set
+            val crossed = crossTuned.mapPartition {
+              (iter, out: Collector[(FlinkVector, FlinkVector, Long, Double)]) => {
+                for ((training, testing) <- iter) {
+                  // use a quadtree if (4^dim)Ntest*log(Ntrain)
+                  // < Ntest*Ntrain, and distance is Euclidean
+                  val useQuadTree = resultParameters.get(UseQuadTreeParam).getOrElse(
+                    math.log(4.0) * training.values.head.size + math.log(math.log(training.values.length))
```

Comment:
Travis reports this line exceeds maximum characters limit (100 characters per line).
https://s3.amazonaws.com/archive.travis-ci.org/jobs/131163461/log.txt